File: compiler/python/state-buffer-estimate/arch/arch.py
from .psumbuffer         import PsumBuffer
from .pearray            import PeArray
from .statebuffer        import StateBuffer
from .poolingeng         import PoolingEng
from .activationeng      import ActivationEng

##########################################################
class Arch(object):

    #-----------------------------------------------------------------
    def __init__(self):

        numberPeRows            = 128
        numberPeColumns         = 64
        numberPsumBanks         = 4
        numberPsumBankEntries   = 256
        sbPartitionsSize        = 12 * 1024 * 1024 // numberPeRows  ## 12 MB
        sbPartitionsSize        =  8 * 1024 * 1024 // numberPeRows  ##  8 MB

        self.__PeArray          = PeArray(numberPeRows, numberPeColumns)
        self.__PsumBuffer       = PsumBuffer(self.gPeArray(), numberPsumBanks,
                                             numberPsumBankEntries)
        self.__PoolingEng       = PoolingEng(self.gPsumBuffer())
        self.__ActivationEng    = ActivationEng(self.gPsumBuffer())
        self.__StateBuffer      = StateBuffer(self.gPeArray(), sbPartitionsSize)

    #-----------------------------------------------------------------
    def gPeArray(self):
        return self.__PeArray

    #-----------------------------------------------------------------
    def gStateBuffer(self):
        return self.__StateBuffer

    #-----------------------------------------------------------------
    def gPsumBuffer(self):
        return self.__PsumBuffer

    #-----------------------------------------------------------------
    def gPoolingEng(self):
        return self.__PoolingEng

    #-----------------------------------------------------------------
    def gActivationEng(self):
        return self.__ActivationEng


    #-----------------------------------------------------------------
    def gNumberPeArrayRows(self):
        return self.__PeArray.gNumberRows()

    #-----------------------------------------------------------------
    def gNumberPeArrayColumns(self):
        return self.__PeArray.gNumberColumns()



    #-----------------------------------------------------------------
    def gNumberPsumBanks(self):
        return self.gPsumBuffer().gNumberBanks()

    #-----------------------------------------------------------------
    def gPsumBankEntries(self):
        return self.gPsumBuffer().gNumberBankEntries()

File: compiler/kcc/arch/src/arch.cpp
#include "poolingeng.hpp"
#include "activationeng.hpp"
#include "statebuffer.hpp"

#include "arch.hpp"

namespace kcc {
namespace arch {

//--------------------------------------------------------
Arch::Arch()
{
    const int  numberPeRows            = 128;
    const int  numberPeColumns       = 64;
    const int  numberPsumBanks         = 4;
    const int  numberPsumBankEntries   = 256;

    //const long stateBuffersSizeInBytes = 12 * 1024 * 1024; // numberPeRows  ## 12 MB
    const long stateBuffersSizeInBytes =  8 * 1024 * 1024; // numberPeRows  ##  8 MB

    const long sbPartitionSizeInBytes  = stateBuffersSizeInBytes  / numberPeRows;

    m_PeArray          = new PeArray(numberPeRows, numberPeColumns);  // first

    m_PsumBuffer       = new PsumBuffer(gPeArray(), numberPsumBanks,
                                         numberPsumBankEntries);
    m_PoolingEng       = new PoolingEng(gPsumBuffer());
    m_ActivationEng    = new ActivationEng(gPsumBuffer());
    m_StateBuffer      = new StateBuffer(gPeArray(), sbPartitionSizeInBytes);
}

//----------------------------------------------------------------
int Arch::gNumberPeArrayRows() const
{
    return m_PeArray->gNumberRows();
}

//----------------------------------------------------------------
int Arch::gNumberPeArrayColumns() const
{
    return m_PeArray->gNumberColumns();
}



//----------------------------------------------------------------
int Arch::gNumberPsumBanks() const
{
    return m_PsumBuffer->gNumberBanks();
}

//----------------------------------------------------------------
int Arch::gPsumBankEntries() const
{
    return m_PsumBuffer->gNumberBankEntries();
}

//----------------------------------------------------------------
const std::string&
Arch::gArchVersion() const
{
    static const std::string version("Tonga-0.2");
    return version;
}

}}


